src/circuit_proof/mod.rs

```diff
@@ -0,0 +1,781 @@
+            // l_poly.2 = a_O
+            l_poly.2[i] = a_O[i];
+            // l_poly.3 = s_L
+            l_poly.3[i] = s_L[i];
+            // r_poly.0 = (z * z^Q * W_O) - y^n
+            r_poly.0[i] = z_zQ_WO[i] - exp_y;
+            // r_poly.1 = y^n * a_R + (z * z^Q * W_L)
+            r_poly.1[i] = exp_y * a_R[i] + z_zQ_WL[i];
+            // r_poly.2 = 0
+            // r_poly.3 = y^n * s_R
+            r_poly.3[i] = exp_y * s_R[i];
+
+            exp_y = exp_y * y; // y^i -> y^(i+1)
+            exp_y_inv = exp_y_inv * y_inv; // y^-i -> y^-(i+1)
+        }
+
+        let t_poly = l_poly.inner_product(&r_poly);
+
+        let t_1_blinding = Scalar::random(rng);
+        let t_3_blinding = Scalar::random(rng);
+        let t_4_blinding = Scalar::random(rng);
+        let t_5_blinding = Scalar::random(rng);
+        let t_6_blinding = Scalar::random(rng);
+
+        let T_1 = gen.pedersen_generators.commit(t_poly.t1, t_1_blinding);
+        let T_3 = gen.pedersen_generators.commit(t_poly.t3, t_3_blinding);
+        let T_4 = gen.pedersen_generators.commit(t_poly.t4, t_4_blinding);
+        let T_5 = gen.pedersen_generators.commit(t_poly.t5, t_5_blinding);
+        let T_6 = gen.pedersen_generators.commit(t_poly.t6, t_6_blinding);
+
+        transcript.commit(T_1.compress().as_bytes());
+        transcript.commit(T_3.compress().as_bytes());
+        transcript.commit(T_4.compress().as_bytes());
+        transcript.commit(T_5.compress().as_bytes());
+        transcript.commit(T_6.compress().as_bytes());
+        let x = transcript.challenge_scalar();
+
+        // t_2_blinding = <z*z^Q, W_V * v_blinding>
+        // in the t_x_blinding calculations, line 76.
+        let t_2_blinding = circuit
+            .W_V
+            .iter()
+            .zip(util::exp_iter(z))
+            .map(|(W_V_i, exp_z)| z * exp_z * inner_product(&W_V_i, &v_blinding))
+            .sum();
+
+        let t_blinding_poly = util::Poly6 {
+            t1: t_1_blinding,
+            t2: t_2_blinding,
+            t3: t_3_blinding,
+            t4: t_4_blinding,
+            t5: t_5_blinding,
+            t6: t_6_blinding,
+        };
+
+        let t_x = t_poly.eval(x);
+        let t_x_blinding = t_blinding_poly.eval(x);
+        let l_vec = l_poly.eval(x);
+        let r_vec = r_poly.eval(x);
+        let e_blinding = x * (i_blinding + x * (o_blinding + x * s_blinding));
+
+        transcript.commit(t_x.as_bytes());
+        transcript.commit(t_x_blinding.as_bytes());
+        transcript.commit(e_blinding.as_bytes());
+
+        // Get a challenge value to combine statements for the IPP
+        let w = transcript.challenge_scalar();
+        let Q = w * gen.pedersen_generators.B;
+
+        let ipp_proof = InnerProductProof::create(
+            transcript,
+            &Q,
+            util::exp_iter(y.invert()),
+            gen.G.to_vec(),
+            gen.H.to_vec(),
+            l_vec.clone(),
+            r_vec.clone(),
```

Comment:
The existing `InnerProductProof` code is expecting that the vector lengths are a power of two, but this isn't the case any more in the circuit code.  How do we want to handle non-power-of-two inner products?